### PR TITLE
ARC-1996 Incremental Backfill

### PR DESCRIPTION
**Pull request by Harminder84**
**What's in this PR?**
Implemented incremental backfilling for commits

**Why**
Allow large customers to backfill their data in chunks rather than in single go. 

**Added feature flags**
USE_BACKFILL_ALGORITHM_INCREMENTAL

**Affected issues**  
[ARC-1996]

**How has this been tested?**  
Local

[ARC-1996]: https://softwareteams.atlassian.net/browse/ARC-1996

### CODE REVIEW

File: src/routes/api/installation/api-installation-sync-post.test.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ describe("ApiInstallationSyncPost", ()=>{
 			expect(findOrStartSync).toBeCalledWith(expect.objectContaining({
 				gitHubInstallationId: GHES_GITHUB_INSTALLATION_ID,
 				gitHubAppId: GHES_GITHUB_APP_ID
-			}), expect.anything(), false, undefined);
+			}), expect.anything(), undefined);
```
Comment (joshkay10) [inline src/routes/api/installation/api-installation-sync-post.test.ts:35]: /nit can drop the `, undefined` now

File: src/routes/jira/sync/jira-sync-post.ts [hunk 2]
```diff
@@ -29,7 +30,13 @@ export const JiraSyncPost = async (req: Request, res: Response, next: NextFuncti
 			res.status(400).send("Invalid date value, cannot select a future date!");
 			return;
 		}
-		await findOrStartSync(subscription, req.log, false, syncType, commitsFromDate);
+
+		const shouldUseBackfillAlgoIncremental = await booleanFlag(BooleanFlags.USE_BACKFILL_ALGORITHM_INCREMENTAL, res.locals.installation.jiraHost);
+		if (shouldUseBackfillAlgoIncremental && isIncrementalBackfilling(subscription, commitsFromDate)) {
+			await findOrStartSync(subscription, req.log, "partial", commitsFromDate, ["pull", "branch", "commit", "build", "deployment"]);
```
Comment (joshkay10) [inline src/routes/jira/sync/jira-sync-post.ts:36]: if you send no tasks it does all tasks so you can delete `["pull", "branch", "commit", "build", "deployment"]`

File: src/routes/jira/sync/jira-sync-post.ts [hunk 3]
```diff
@@ -59,3 +66,13 @@ const getStartTimeInDaysAgo = (commitsFromDate: Date | undefined) => {
...
 };
+
+const isIncrementalBackfilling = (subscription: Subscription, fromDate: Date | undefined): boolean => {
+	if (subscription.syncStatus === "FAILED") {
+		return false;
+	}
+	if (!fromDate) {
+		return false;
+	}
+	return true;
```
Comment (joshkay10) [inline src/routes/jira/sync/jira-sync-post.ts:77]: /nit
could replace line 74-77 with `return Boolean(fromDate);`

File: src/models/reposyncstate.ts [hunk 4]
```diff
@@ -34,6 +34,8 @@ export class RepoSyncState extends Model {
 	config?: Config;
 	updatedAt: Date;
 	createdAt: Date;
+	commitFrom: Date;
```
Comment (gxueatlassian) [inline src/models/reposyncstate.ts:37]: Shouldn't it be `commitFrom?: Date`

File: src/models/reposyncstate.ts [hunk 5]
```diff
@@ -210,5 +213,6 @@ RepoSyncState.init({
 	syncCompletedAt: DATE,
 	config: JSON,
 	createdAt: DATE,
-	updatedAt: DATE
+	updatedAt: DATE,
+	commitFrom: DATE
```
Comment (gxueatlassian) [inline src/models/reposyncstate.ts:217]: /nit just for code readability, maybe move this line above to after the `deploymentCursor`, so it is like
```
xxxStatus
xxxStatus
xxxStatus
xxxCursor
xxxCursor
xxxCursor
xxxFrom
xxxFrom
xxxFrom
```

File: src/routes/jira/sync/jira-sync-post.ts (hunk 2 above)
Comment (gxueatlassian) [reply src/routes/jira/sync/jira-sync-post.ts:36]: +1
I think I had a discussion with @Harminder84 before and thought we need those. 
But now with a second thought, I don't remember why do we need those. Do you remember @Harminder84 ?

File: src/routes/jira/sync/jira-sync-post.ts [hunk 6]
```diff
@@ -59,3 +66,13 @@ const getStartTimeInDaysAgo = (commitsFromDate: Date | undefined) => {
 	if (commitsFromDate === undefined) return undefined;
 	return Math.floor((Date.now() -  commitsFromDate?.getTime()) / MILLISECONDS_IN_ONE_DAY);
 };
+
+const isIncrementalBackfilling = (subscription: Subscription, fromDate: Date | undefined): boolean => {
+	if (subscription.syncStatus === "FAILED") {
+		return false;
+	}
+	if (!fromDate) {
+		return false;
```
Comment (gxueatlassian) [inline src/routes/jira/sync/jira-sync-post.ts:75]: /nit
not sure whether we should do it in this PR or another, probably should check the param and always make sure the `commitsFrom` is not empty.

File: src/sync/installation.ts [hunk 7]
```diff
@@ -110,7 +110,19 @@ export const updateJobStatus = async (
 	const status = isComplete ? "complete" : "pending";
 
 	logger.info({ status }, "Updating job status");
-	await updateRepo(subscription, repositoryId, { [getStatusKey(task)]: status });
+
+	const values: { [x: string]: string | Date} = { [getStatusKey(task)]: status };
```
Comment (gxueatlassian) [inline src/sync/installation.ts:114]: I know it is hard, can we give a better name than `value`?

File: src/sync/installation.ts [hunk 8]
```diff
@@ -110,7 +110,19 @@ export const updateJobStatus = async (
...
+
+	if (isComplete && task === "commit" && data.commitsFromDate) {
+		const repoSync = await RepoSyncState.findByRepoId(subscription, repositoryId);
+		const commitsFromDate =  new Date(data.commitsFromDate);
+		// Set commitsFromDate in RepoSyncState only if its the later date
+		if (repoSync &&	(!repoSync.commitFrom || repoSync.commitFrom.getTime() > commitsFromDate.getTime())) {
```
Comment (gxueatlassian) [inline src/sync/installation.ts:120]: maybe this is easier to read?
```
const newCommitsFromDate =  new Date(data.commitsFromDate);
(repoSync && repoSync.commitFrom && repoSync.commitFrom.getTime() > newCommitsFromDate.getTime()
```

File: src/sync/installation.ts [hunk 9]
```diff
@@ -110,7 +110,19 @@ export const updateJobStatus = async (
 	const status = isComplete ? "complete" : "pending";
 
 	logger.info({ status }, "Updating job status");
-	await updateRepo(subscription, repositoryId, { [getStatusKey(task)]: status });
+
+	const values: { [x: string]: string | Date} = { [getStatusKey(task)]: status };
+
+	if (isComplete && task === "commit" && data.commitsFromDate) {
```
Comment (gxueatlassian) [inline src/sync/installation.ts:116]: /nit
Just to clarify, once we implement `commitFrom` for all tasks, this `if ()` branch will be generic, right?

File: src/sync/installation.ts (hunk 8 above)
Comment (Harminder84) [reply src/sync/installation.ts:120]: One case missing in your expression when repoSync.commitFrom is null

File: src/sync/installation.ts (hunk 9 above)
Comment (Harminder84) [reply src/sync/installation.ts:116]: Yep, then we can remove task condition in above expression

File: src/routes/jira/sync/jira-sync-post.ts (hunk 2 above)
Comment (Harminder84) [reply src/routes/jira/sync/jira-sync-post.ts:36]: At some places we're relying on tasks to clear data. Once instance is 

```
const resetTargetedTasks = async (subscription: Subscription, syncType?: SyncType, targetTasks?: TaskType[]): Promise<void> => {
	if (!targetTasks?.length) {
		return;
	}

```

Comment (gxueatlassian) [reply src/routes/jira/sync/jira-sync-post.ts:40]: Okay, I see the reason `why` now. 
That `if (!targetTasks?.length) {	return;	}` doesn't make sense to me, but I'd agree we should leave it alone for now.

File: src/models/reposyncstate.ts [hunk 10]
```diff
@@ -160,8 +160,7 @@ export class RepoSyncState extends Model {
 			buildStatus: null,
 			buildCursor: null,
 			deploymentStatus: null,
-			deploymentCursor: null,
-			commitFrom: null
```
Comment (gxueatlassian) [inline src/models/reposyncstate.ts]: Hmm why remove it? `commitFrom: null`

File: src/routes/jira/sync/jira-sync-post.ts [hunk 11]
```diff
@@ -9,8 +9,12 @@ import { booleanFlag, BooleanFlags } from "~/src/config/feature-flags";
 export const JiraSyncPost = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
 	const { installationId: gitHubInstallationId, appId: gitHubAppId } = req.body;
 
+	if (!req.body?.commitsFromDate) {
+		res.status(404).send("Missing sync date");
```
Comment (gxueatlassian) [inline src/routes/jira/sync/jira-sync-post.ts:13]: `400` instead of `404`?

File: src/routes/jira/sync/jira-sync-post.ts [hunk 12]
```diff
@@ -9,8 +9,12 @@ import { booleanFlag, BooleanFlags } from "~/src/config/feature-flags";
...
+		return;
+	}
 	// A date to start fetching commit history(main and branch) from.
-	const commitsFromDate = req.body.commitsFromDate ? new Date(req.body.commitsFromDate) : undefined;
+	const commitsFromDate = new Date(req.body.commitsFromDate);
```
Comment (gxueatlassian) [inline src/routes/jira/sync/jira-sync-post.ts:17]: /nit
Ideally the format of the `req.body.commitsFrom` will need validation as well.

File: src/routes/jira/sync/jira-sync-post.ts (hunk 11 above)
Comment (Harminder84) [reply src/routes/jira/sync/jira-sync-post.ts]: Again made it optional. There are few scenarios where this API is being called without `commitFromDate`

File: src/routes/jira/sync/jira-sync-post.ts (hunk 12 above)
Comment (Harminder84) [reply src/routes/jira/sync/jira-sync-post.ts]: Date format validation is already done at router level.

File: src/routes/jira/sync/jira-sync-post.ts [hunk 13]
```diff
@@ -29,7 +31,13 @@ export const JiraSyncPost = async (req: Request, res: Response, next: NextFuncti
 			res.status(400).send("Invalid date value, cannot select a future date!");
 			return;
 		}
-		await findOrStartSync(subscription, req.log, false, syncType, commitsFromDate);
+
+		const shouldUseBackfillAlgoIncremental = await booleanFlag(BooleanFlags.USE_BACKFILL_ALGORITHM_INCREMENTAL, res.locals.installation.jiraHost);
+		if (shouldUseBackfillAlgoIncremental && isIncrementalBackfilling(subscription, syncType, commitsFromDate)) {
+			await findOrStartSync(subscription, req.log, syncType, commitsFromDate, ["pull", "branch", "commit", "build", "deployment"]);
```
Comment (gxueatlassian) [inline src/routes/jira/sync/jira-sync-post.ts:37]: Isn't it partial for the syncType here?

File: static/js/jira-configuration.js [hunk 14]
```diff
@@ -89,7 +89,7 @@ $(".sync-connection-link").click(event => {
 		event.preventDefault();
 		const commitsFromDate = document.getElementById('backfill-date-picker').value;
 		window.AP.context.getToken(function (jwt) {
-			restartBackfillPost({jwt, _csrf: csrfToken, jiraHost, syncType: "full", installationId, commitsFromDate, appId});
+			restartBackfillPost({jwt, _csrf: csrfToken, jiraHost, syncType: "partial", installationId, commitsFromDate, appId});
```
Comment (gxueatlassian) [inline static/js/jira-configuration.js:92]: Okay, I got it. But this isn't controlled by the FF. Maybe do it on the WebServer side.

Comment (Harminder84) [reply static/js/jira-configuration.js]: Put behind FF

**Review by gxueatlassian — APPROVED** [review]

**Pull request merged**